[doc] Add the CW340 to the get started guide (#19691)
lowRISC/opentitan: doc/getting_started/setup_fpga.md

## Obtain an FPGA bitstream

To run OpenTitan on an FPGA, you will need an FPGA bitstream.

> You can either download the latest bitstream for the ChipWhisperer board or build it yourself.
> For the purpose of this tutorial will set the board model in a environment variable but feel free to inline the board model into the commands bellow.

Reviewer: ```suggestion
To instantiate OpenTitan on an FPGA you will need a bitstream. You can either download an existing bitstream for one of the supported ChipWhisperer boards or you can build it yourself.
To ensure the commands shown in this tutorial work for any supported board, we use the `BOARD` environment variable to specify the target board.
```

Author: Done

During the build process, the boot ROM is baked into the bitstream.
As mentioned above, we maintain two boot ROM programs, one for testing (_test ROM_), and one for production (_ROM_).

To build an FPGA bitstream with the _test ROM_ for the CW340 board, use:
```sh
cd $REPO_TOP
bazel build //hw/bitstream/vivado:fpga_${BOARD}_test_rom
```

> To build an FPGA bitstream with the _ROM_ for the CW340 board, use:

Reviewer: Explain why, even though we're using a generic `$BOARD` variable, the ROM is for the CW340?

```sh
cd $REPO_TOP
bazel build //hw/bitstream/vivado:fpga_${BOARD}_rom
```

> **Note**: Building these bitstreams will require Vivado be installed on your system, with access to the proper licenses, described [here](./install_vivado/README.md).
> For general software development on the FPGA, Vivado must still be installed, but the Lab Edition is sufficient.

Reviewer 2: ```suggestion
> **Note**: Building these bitstreams will require Vivado be installed on your system, with access to the proper licenses, described [here](./install_vivado/README.md).
> For general software development on the FPGA, Vivado must still be installed, but the Lab Edition is sufficient.
```

Quoting this may be nice to indicate a different voice.

Author: Neat, done

source lowrisc_systems_chip_earlgrey_${BOARD}_0.1.tcl
```

## Connecting the ChipWhisperer FPGA board:

### CW310 board

The CW310 board supports different power options.
It is recommended to power the board via the included DC power adapter.
To this end:
1. Set the *SW2* switch (to the right of the barrel connector) up to the *5V Regulator* option.
2. Set the switch below the barrel connector to the right towards the *Barrel* option.
3. Set the *Control Power* switch (bottom left corner, *SW7*) to the right.
4. Ensure the *Tgt Power* switch (above the fan, *S1*) is set to the right towards the *Auto* option.
5. Plug the DC power adapter into the barrel jack (*J11*) in the top left corner of the board.
6. Use a USB-C cable to connect your PC with the *USB-C Data* connector (*J8*) in the lower left corner on the board.
### CW340 board

The CW340 board should be powered via the included DC power adapter.
To this end:
1. Set the *Control Power* switch (top left corner, *SW7*) to the right.
2. Ensure the *Tgt Power* switch (center of the board) is set to the right towards the *Auto* option.
3. Plug the DC power adapter into the barrel jack (*J11*) in the top left corner of the board.

> 4. Use a USB-C cable to connect your PC (*host*) with the *USB-C* connector (*J28*) in the lower left corner on the board.

Reviewer 3: This guide show say something about the serial switches (JP1-4). This impacts the name of the serial devices below.

Author: done

### Detecting the board
You can now use the following to monitor output from dmesg:
```sh
sudo dmesg -Hw
```
This should show which serial ports have been assigned, or if the board is having trouble connecting to USB.
If `dmesg` reports a problem you can trigger a *USB_RST* with *SW5*.
When properly connected, `dmesg` should identify the board, not show any errors, and the status light should flash (next to *SW5*).

> They should be named `'/dev/ttyACM*'` for CW310 and `/dev/ttyUSB*` + `/dev/ttyACM*` for CW340, e.g. `/dev/ttyACM1`.

Reviewer 3: This is not quite correct: it depends on the jumpers for serial. If jumpers are all on SAM3x (JP1-4) then it should be `/dev/ttyACM*`. If it's the FTDI, it should be `/dev/ttyUSB*`

Author: Now I'm mentioning the jumpers.

But during debugging this behavior is not helpful.
The `--no-export` option of FuseSoC disables copying the source files into the staging area, and `--setup` instructs fusesoc to not start the synthesis process.

**Only create Vivado project directory by using FuseSoC directly (skipping Bazel invocation).**
```sh
cd $REPO_TOP
fusesoc --cores-root . run --flag=fileset_top --target=synth --no-export --setup lowrisc:systems:chip_earlgrey_${BOARD}
```

You can then navigate to the created project directory, and open Vivado
```sh
. /tools/Xilinx/Vivado/{{#tool-version vivado }}/settings64.sh
cd $REPO_TOP/build/lowrisc_systems_chip_earlgrey_${BOARD}_0.1/synth-vivado/vivado
```

Finally, using the Tcl console, you can kick off the project setup with
```sh
source lowrisc_systems_chip_earlgrey_${BOARD}_0.1.tcl
```

## Connecting the ChipWhisperer FPGA board:

### CW310 board

The CW310 board supports different power options.
It is recommended to power the board via the included DC power adapter.
To this end:
1. Set the *SW2* switch (to the right of the barrel connector) up to the *5V Regulator* option.

> 2. Set the switch below the barrel connector to the right towards the *Barrel* option.

Reviewer: ```suggestion
2. Set the switch below the barrel connector to the right, towards the *Barrel* option.
```

3. Set the *Control Power* switch (bottom left corner, *SW7*) to the right.

> 4. Ensure the *Tgt Power* switch (above the fan, *S1*) is set to the right towards the *Auto* option.

Reviewer: ```suggestion
4. Ensure the *Tgt Power* switch (above the fan, *S1*) is set to the right, towards the *Auto* option.
```

5. Plug the DC power adapter into the barrel jack (*J11*) in the top left corner of the board.
6. Use a USB-C cable to connect your PC with the *USB-C Data* connector (*J8*) in the lower left corner on the board.

### CW340 board

The CW340 board should be powered via the included DC power adapter.
To this end:
1. Set the *Control Power* switch (top left corner, *SW7*) to the right.

> 1. Ensure the *Tgt Power* switch (center of the board) is set to the right towards the *Auto* option.

Reviewer: ```suggestion
1. Ensure the *Tgt Power* switch (center of the board) is set to the right, towards the *Auto* option.
```

1. Plug the DC power adapter into the barrel jack (*J11*) in the top left corner of the board.
1. Use a USB-C cable to connect your PC (*host*) with the *USB-C* connector (*J28*) in the lower left corner on the board.
1. Set the jumpers *JP1* and *JP2* to select the UART0 routing:
   1. If set to FTDI the UART0 will (likely) be routed to `/dev/ttyUSB2`.

>    1. If set to SAM the UART0 will be routed to `/dev/ttyACM0`.

Reviewer 2: Could we number these to make them look nicer in plain text?

Author: Yeah, I just don't know how to make the sub items without adding two tailing spaces on the previous line.

```sh
bazel run //sw/host/opentitantool -- \
--exec "bootstrap $(ci/scripts/target-location.sh //sw/device/examples/hello_world:hello_world_fpga_${BOARD}_bin)"\
console
```

#### Troubleshooting

If the firmware load fails, try pressing the "USR-RST" button before loading the bitstream.

## Debugging with JTAG

### Tap Straps

After bootstrapping the firmware, the TAP straps may need to be set.
As of this writing, the FPGA images are typically programmed to be in the RMA lifecycle state, and the TAP straps are sampled continuously in that state.
To connect the JTAG chain to the CPU's TAP, adjust the strap values with opentitantool.
Assuming opentitantool has been built and that the current directory is the root of the workspace, run these commands:

```sh
./bazel-bin/sw/host/opentitantool/opentitantool --interface cw340 \
        --exec "gpio write TAP_STRAP0 false" \
        --exec "gpio write TAP_STRAP1 true" \
        no-op
```

### CW340 Board

The CW340 supports JTAG-based debugging with OpenOCD and GDB via the FTDI Chip embedded on the board. So no external JTAG adapter is needed.

> ### CW310 Board

Reviewer 4: ```suggestion

### CW310 Board
```

1. Observe the output both on the board and the serial console. Type any text into the console window.
1. Exit `screen` by pressing CTRL-a k, and confirm with y.

> 1. Alternatively you can use the console embedded into the opentitantool to see the output from the UART instead of `screen`.

Reviewer: ```suggestion
1. Alternatively, you can use the console embedded into `opentitantool` to see the output from the UART instead of `screen`.
```
